### PR TITLE
Add experimental PyTorch MNIST TPU pod test

### PR DESCRIPTION
**Pull request by will-cromar**

### CODE REVIEW

**Comment by will-cromar** [comment]
To test this out, run `jsonnet templates/oneshot.jsonnet -S --tla-str test=pt-nightly-mnist-pods-experimental-pods | kubectl create -f -`

**Comment by jysohn23** [comment]
Nice! Thanks 😄

**Comment by will-cromar** [comment]
Updates since I published this draft:

- Use TPU provided by Kubernetes, which lets us queue jobs when no quota is available.
- Create base template for pods tests that hides the complexity of managing instance groups from the user and exposes the same interface as the other template.
- Use the smaller `google/cloud-sdk:slim` image as a base image for PyTorch pods tests, since we only use `gcloud` on the main K8s Pod anyway.
- Explicitly wait for instance group to initialize in setup script.
- Lower timeout for setup script since the TPU is already READY when the K8s pod starts.

**Comment by will-cromar** [comment]
Tested: [oneshot link](https://pantheon.corp.google.com/kubernetes/job/europe-west4-a/europe-oneshots/default/pt-nightly-mnist-convergence-v2-32-cmbhv?project=xl-ml-test&tab=details&duration=P30D&pod_summary_list_tablesize=20&service_list_datatablesize=20)

File: images/pytorch-pods/entrypoint.sh [hunk 1]
```diff
@@ -13,24 +13,36 @@
...
+export master=$(gcloud compute instance-groups \
+  list-instances \
+  ${INSTANCE_GROUP_NAME} \
+  --zone=${ZONE} \
+  --format="value(NAME)" \
+  --limit=1) && \
+gcloud -q compute ssh --internal-ip --zone=$ZONE $master \
+  --command "source /anaconda3/etc/profile.d/conda.sh && \
+  conda activate torch-xla-nightly && \
+  python -m torch_xla.distributed.xla_dist --tpu=$TPU_POD_NAME --conda-env=torch-xla-nightly -- $*"
```
Comment (jysohn23) [inline images/pytorch-pods/entrypoint.sh:44]: This way for PT/TPU Pods tests jsonnet configs, we're only able to specify the `python $SCRIPT $ARGS` without the `xla_dist` module flags? I think we need the ability to specify other flags like `--env` (it's an `xla_dist` module flag) and also switch out the `--conda-env` versions to like `torch-xla-1.5` or smth.

Can we build the `python -m torch_xla...` full command from jsonnet?

**Comment by taylanbil** [comment]
Awesome!

File: images/pytorch-pods/entrypoint.sh [hunk 2]
```diff
@@ -13,24 +13,36 @@
...
+mkdir -p /root/google-cloud-sdk/bin/
+ln -s $(which gcloud) /root/google-cloud-sdk/bin/gcloud
 source /publish.sh
 
-# "$@" should look like python -m torch_xla.distributed.xla_dist --tpu=...
-docker-entrypoint.sh "$@"
-
-# Teardown resources
-/teardown-pytorch-pods.sh
+timeout 180 /setup-instances.sh && \
```
Comment (zcain117) [inline images/pytorch-pods/entrypoint.sh:34]: If `setup-instances.sh` fails, will the overall Job fail? And would the logs from `entrypoint.sh`, `setup-instances.sh`, and `teardown-instances.sh` appear in stackdriver?

Comment (will-cromar) [reply images/pytorch-pods/entrypoint.sh:34]: Yes and yes. I save the exit code from this chain of commands, unconditionally tear down any resources, and then return that exit code.

File: images/pytorch-pods/entrypoint.sh (hunk 1 above)
Comment (will-cromar) [reply images/pytorch-pods/entrypoint.sh]: Added template parameters for conda env and `xla_dist` flags. [oneshot link](https://pantheon.corp.google.com/kubernetes/job/europe-west4-a/europe-oneshots/default/pt-nightly-mnist-convergence-v2-32-nfnsj?project=xl-ml-test&tab=details&duration=P30D&pod_summary_list_tablesize=20&service_list_datatablesize=20)

**Pull request merged**